contracts/core/liquidators/TWAPLiquidator.sol

```diff
@@ -0,0 +1,291 @@
+ */
+contract TWAPLiquidator is
+    ILiquidator,
+    TWAPAuction,
+    AuctionGetters,
+    TWAPAuctionGetters,
+    Ownable
+{
+    using SafeMath for uint256;
+
+    ICore public core;
+    string public name;
+    mapping(address => TWAPAuction.TWAPState) public auctions;
+
+    /* ============ Modifier ============ */
+    modifier isValidSet() {
+        requireValidSet(msg.sender);
+        _;
+    }
+
+    /**
+     * TWAPLiquidator constructor
+     *
+     * @param _core                   Core instance
+     * @param _oracleWhiteList        Oracle WhiteList instance
+     * @param _auctionPeriod          Length of auction
+     * @param _rangeStart             Percentage above FairValue to begin auction at
+     * @param _rangeEnd               Percentage below FairValue to end auction at     
+     * @param _assetPairHashes        List of asset pair unique identifiers
+     * @param _assetPairBounds        List of asset pair USD-denominated chunk auction size bounds
+     * @param _name                   Descriptive name of Liquidator
+     */
+    constructor(
+        ICore _core,
+        IOracleWhiteList _oracleWhiteList,
+        uint256 _auctionPeriod,
+        uint256 _rangeStart,
+        uint256 _rangeEnd,
+        bytes32[] memory _assetPairHashes,
+        BoundsLibrary.Bounds[] memory _assetPairBounds,
+        string memory _name
+    )
+        public
+        TWAPAuction(
+            _oracleWhiteList,
+            _auctionPeriod,
+            _rangeStart,
+            _rangeEnd,
+            _assetPairHashes,
+            _assetPairBounds
+        )
+    {
+        core = _core;
+        name = _name;
+    }
+
+    /* ============ External Functions ============ */
+
+    /**
+     * Initiates a TWAP auction. Can only be called by a SetToken.
+     *
+     * @param _currentSet                   The Set to rebalance from
+     * @param _nextSet                      The Set to rebalance to
+     * @param _startingCurrentSetQuantity   The currentSet quantity to rebalance
+     * @param _liquidatorData               Bytecode formatted data with TWAPLiquidator-specific arguments
+     */
+    function startRebalance(
+        ISetToken _currentSet,
+        ISetToken _nextSet,
+        uint256 _startingCurrentSetQuantity,
+        bytes calldata _liquidatorData
+    )
+        external
+        isValidSet
+    {
+        // Retrieve the chunk auction size and auction period from liquidator data.
+        TWAPAuction.TWAPLiquidatorData memory twapLiquidatorData = TWAPAuction.parseLiquidatorData(_liquidatorData);
+
+        // Validates chunk auction size and chunkAuctionPeriod validity
+        TWAPAuction.validateLiquidatorData(
+            _currentSet,
+            _nextSet,
+            _startingCurrentSetQuantity,
+            twapLiquidatorData
+        );
+
+        // Validates a valid linear auction can be initiated
+        TwoAssetPriceBoundedLinearAuction.validateTwoAssetPriceBoundedAuction(
```

Comment:
Feels like this should be the first thing run. It has the most basic checks and isn't reliant on any previous logic.

Reply:
+1